core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java

```diff
@@ -379,12 +387,27 @@ public void start()
                     return;
                 }
 
+                planningThread = currentThread();
+                stateMachine.getStateChange(PLANNING).addListener(() -> {
+                    if (stateMachine.getQueryState() == FAILED) {
+                        synchronized (this) {
+                            if (planningThread != null) {
+                                planningThread.interrupt();
+                            }
+                        }
+                    }
+                }, directExecutor());
+
                 PlanRoot plan = planQuery();
                 // DynamicFilterService needs plan for query to be registered.
                 // Query should be registered before dynamic filter suppliers are requested in distribution planning.
                 registerDynamicFilteringQuery(plan);
                 planDistribution(plan);
 
+                synchronized (this) {
+                    planningThread = null;
```

Comment:
you need to unset interrupted flag here too. Fortunately `Threadh.interrupt` and `Thread.interrupted` are in happens-before relationship